allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -1,18 +1,82 @@
-from allennlp.data.dataset import InstanceCollection
+from typing import Iterable, Iterator, Callable
+
 from allennlp.data.instance import Instance
 from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
 from allennlp.common.registrable import Registrable
+from allennlp.common.util import ensure_list
+
+class _LazyInstances(Iterable):
+    """
+    An ``Iterable`` that just wraps a thunk for generating instances and calls it for
+    each call to ``__iter__``.
+    """
+    def __init__(self, instance_generator: Callable[[], Iterator[Instance]]) -> None:
+        super().__init__()
+        self.instance_generator = instance_generator
 
+    def __iter__(self) -> Iterator[Instance]:
+        instances = self.instance_generator()
+        if isinstance(instances, list):
+            raise ConfigurationError("For a lazy dataset reader, _read() must return a generator")
+        return instances
 
 class DatasetReader(Registrable):
     """
-    A ``DatasetReader`` reads data from some location and constructs a :class:`Dataset`.  All
-    parameters necessary to read the data apart from the filepath should be passed to the
-    constructor of the ``DatasetReader``.
+    A ``DatasetReader`` knows how to turn a file containing a dataset into a collection
+    of ``Instance`` s.  To implement your own, just override the `_read(file_path)` method
+    to return an ``Iterable`` of the instances. This could be a list containing the instances
+    or a lazy generator that returns them one at a time.
+
+    All parameters necessary to _read the data apart from the filepath should be passed
+    to the constructor of the ``DatasetReader``.
+
+    Parameters
+    ----------
+    lazy : ``bool``, optional (default=False)
+        If this is true, ``instances()`` will return an object whose ``__iter__`` method
+        reloads the dataset each time it's called. Otherwise, ``instances()`` returns a list.
     """
-    def read(self, file_path: str) -> InstanceCollection:
+    def __init__(self, lazy: bool = False) -> None:
+        self.lazy = lazy
+
+    def read(self, file_path: str) -> Iterable[Instance]:
+        """
+        Returns an ``Iterable`` containing all the instances
+        in the specified dataset.
+
+        If ``self.lazy`` is False, this calls ``self._read()``,
+        calls ``ensure_list`` on the result (which is a no-op if
+        self._read() returns a list, which most of the existing
+        dataset readers do), and returns the resulting list.
+
+        If ``self.lazy`` is True, this returns an object whose
+        ``__iter__`` method calls ``self._read()`` each iteration.
+        Note that if your implementation of ``_read()`` is not lazy
+        (i.e. it loads all instances into memory at once), then your
+        "lazy" dataset reader is just loading the entire dataset into
+        a list each time you iterate over it, which is probably not
```

Comment:
You crash now, so this comment should be updated.

Reply:
done